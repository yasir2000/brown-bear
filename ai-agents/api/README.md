# üåê Brown Bear AI Agents API

REST and GraphQL API server for the AI Multi-Agent System.

## Features

- **REST API**: Comprehensive RESTful endpoints
- **GraphQL API**: Flexible query language for complex operations
- **WebSocket**: Real-time updates and subscriptions
- **Authentication**: JWT-based auth with API keys
- **Rate Limiting**: Protect against abuse
- **API Documentation**: Auto-generated OpenAPI/Swagger docs

## Quick Start

```bash
# Install dependencies
pnpm install

# Configure environment
cp .env.example .env

# Start the server
pnpm dev

# Production build
pnpm build
pnpm start
```

## API Endpoints

### REST API

- `POST /api/v1/agents/create` - Create agent
- `GET /api/v1/agents` - List agents
- `GET /api/v1/agents/:id` - Get agent
- `PUT /api/v1/agents/:id` - Update agent
- `DELETE /api/v1/agents/:id` - Delete agent
- `POST /api/v1/tasks/create` - Create task
- `GET /api/v1/tasks/:id` - Get task
- `GET /api/v1/metrics` - System metrics

### GraphQL API

Access at `http://localhost:8080/graphql`

### WebSocket

Connect at `ws://localhost:8080/ws`

## Documentation

Interactive API documentation available at `http://localhost:8080/docs`
