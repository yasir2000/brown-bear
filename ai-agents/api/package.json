{
  "name": "@brownbear/ai-agents-api",
  "version": "1.0.0",
  "description": "REST and GraphQL API for Brown Bear AI Multi-Agent System",
  "main": "dist/index.js",
  "scripts": {
    "dev": "tsx watch src/index.ts",
    "build": "tsc",
    "start": "node dist/index.js",
    "test": "jest",
    "lint": "eslint src --ext .ts"
  },
  "keywords": ["api", "rest", "graphql", "ai", "agents"],
  "author": "BrownBear Team",
  "license": "Apache-2.0",
  "dependencies": {
    "@brownbear/ai-agents-framework": "workspace:*",
    "express": "^4.18.2",
    "apollo-server-express": "^3.13.0",
    "graphql": "^16.8.1",
    "express-rate-limit": "^7.1.5",
    "helmet": "^7.1.0",
    "cors": "^2.8.5",
    "jsonwebtoken": "^9.0.2",
    "bcryptjs": "^2.4.3",
    "joi": "^17.12.0",
    "ws": "^8.16.0",
    "swagger-ui-express": "^5.0.0",
    "swagger-jsdoc": "^6.2.8",
    "compression": "^1.7.4",
    "morgan": "^1.10.0",
    "dotenv": "^16.4.0"
  },
  "devDependencies": {
    "@types/express": "^4.17.21",
    "@types/node": "^20.11.0",
    "@types/ws": "^8.5.10",
    "@types/cors": "^2.8.17",
    "@types/jsonwebtoken": "^9.0.5",
    "@types/bcryptjs": "^2.4.6",
    "@types/compression": "^1.7.5",
    "@types/morgan": "^1.9.9",
    "@types/swagger-ui-express": "^4.1.6",
    "typescript": "^5.3.3",
    "tsx": "^4.7.0",
    "jest": "^29.7.0",
    "eslint": "^8.56.0"
  }
}
