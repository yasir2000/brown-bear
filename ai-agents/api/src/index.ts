/**
 * Brown Bear AI Agents API Server
 * Express + GraphQL + WebSocket server
 */

import express from 'express';
import { ApolloServer } from 'apollo-server-express';
import { WebSocketServer } from 'ws';
import helmet from 'helmet';
import cors from 'cors';
import compression from 'compression';
import rateLimit from 'express-rate-limit';
import swaggerUi from 'swagger-ui-express';
import swaggerJsdoc from 'swagger-jsdoc';
import dotenv from 'dotenv';
import { AgentFramework } from '@brownbear/ai-agents-framework';
import { setupRESTRoutes } from './routes';
import { typeDefs, resolvers } from './graphql';
import { authenticateToken } from './middleware/auth';

dotenv.config();

const PORT = process.env.PORT || 8080;
const HOST = process.env.HOST || '0.0.0.0';

/**
 * Initialize the API server
 */
async function startServer() {
  const app = express();

  // Middleware
  app.use(helmet());
  app.use(cors());
  app.use(compression());
  app.use(express.json());
  app.use(express.urlencoded({ extended: true }));

  // Rate limiting
  const limiter = rateLimit({
    windowMs: 15 * 60 * 1000, // 15 minutes
    max: 100, // Limit each IP to 100 requests per windowMs
    message: 'Too many requests from this IP',
  });
  app.use('/api/', limiter);

  // Initialize AI Agent Framework
  const framework = new AgentFramework({
    apiKey: process.env.OPENAI_API_KEY,
    orchestratorUrl: process.env.ORCHESTRATOR_URL || 'http://localhost:8080',
    provider: (process.env.AI_PROVIDER as any) || 'openai',
    model: process.env.AI_MODEL || 'gpt-4-turbo',
    redis: {
      host: process.env.REDIS_HOST || 'localhost',
      port: parseInt(process.env.REDIS_PORT || '6379'),
      password: process.env.REDIS_PASSWORD,
    },
    logging: {
      level: (process.env.LOG_LEVEL as any) || 'info',
      format: (process.env.LOG_FORMAT as any) || 'pretty',
    },
  });

  await framework.initialize();

  // Store framework in app context
  app.locals.framework = framework;

  // Health check
  app.get('/health', (req, res) => {
    res.json({
      status: 'healthy',
      timestamp: new Date().toISOString(),
      framework: framework.isInitialized(),
    });
  });

  // REST API routes
  setupRESTRoutes(app);

  // GraphQL setup
  const apolloServer = new ApolloServer({
    typeDefs,
    resolvers,
    context: ({ req }) => ({
      framework,
      user: (req as any).user, // From auth middleware
    }),
  });

  await apolloServer.start();
  apolloServer.applyMiddleware({ app, path: '/graphql' });

  // Swagger API documentation
  const swaggerSpec = swaggerJsdoc({
    definition: {
      openapi: '3.0.0',
      info: {
        title: 'Brown Bear AI Agents API',
        version: '1.0.0',
        description: 'REST and GraphQL API for AI Multi-Agent System',
      },
      servers: [{ url: `http://localhost:${PORT}` }],
    },
    apis: ['./src/routes/*.ts'],
  });

  app.use('/docs', swaggerUi.serve, swaggerUi.setup(swaggerSpec));

  // Start HTTP server
  const server = app.listen(PORT, HOST, () => {
    console.log(`ðŸš€ API Server running at http://${HOST}:${PORT}`);
    console.log(`ðŸ“Š GraphQL endpoint: http://${HOST}:${PORT}/graphql`);
    console.log(`ðŸ“š API docs: http://${HOST}:${PORT}/docs`);
  });

  // WebSocket server for real-time updates
  const wss = new WebSocketServer({ server, path: '/ws' });

  wss.on('connection', (ws) => {
    console.log('WebSocket client connected');

    // Subscribe to framework events
    const handlers = {
      'agent:status': (data: any) => ws.send(JSON.stringify({ type: 'agent:status', data })),
      'task:update': (data: any) => ws.send(JSON.stringify({ type: 'task:update', data })),
      'system:event': (data: any) => ws.send(JSON.stringify({ type: 'system:event', data })),
    };

    Object.entries(handlers).forEach(([event, handler]) => {
      framework.on(event, handler);
    });

    ws.on('close', () => {
      console.log('WebSocket client disconnected');
      // Remove event listeners
      Object.entries(handlers).forEach(([event, handler]) => {
        framework.off(event, handler);
      });
    });

    ws.on('error', (error) => {
      console.error('WebSocket error:', error);
    });
  });

  // Graceful shutdown
  process.on('SIGTERM', async () => {
    console.log('SIGTERM received, shutting down gracefully...');
    server.close(() => {
      console.log('HTTP server closed');
    });
    await framework.shutdown();
    wss.close(() => {
      console.log('WebSocket server closed');
    });
    process.exit(0);
  });
}

// Start the server
startServer().catch((error) => {
  console.error('Failed to start server:', error);
  process.exit(1);
});
