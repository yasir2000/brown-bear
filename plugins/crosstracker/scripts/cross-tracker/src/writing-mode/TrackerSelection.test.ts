/*
 * Copyright (c) Enalean, 2018-Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { shallowMount } from "@vue/test-utils";
import type { Wrapper } from "@vue/test-utils";
import { createCrossTrackerLocalVue } from "../helpers/local-vue-for-test";
import TrackerSelection from "./TrackerSelection.vue";
import * as project_cache from "./projects-cache";
import * as rest_querier from "../api/rest-querier";
import { createStoreMock } from "../../../../../../src/scripts/vue-components/store-wrapper-jest";
import type { Project, SelectedTracker, State, Tracker } from "../type";

describe("TrackerSelection", () => {
    let store = {
        commit: jest.fn(),
    };

    async function instantiateComponent(
        selectedTrackers: Array<SelectedTracker> = []
    ): Promise<Wrapper<TrackerSelection>> {
        const store_options = { state: { is_user_admin: true } as State, commit: jest.fn() };
        store = createStoreMock(store_options);

        return shallowMount(TrackerSelection, {
            localVue: await createCrossTrackerLocalVue(),
            propsData: {
                selectedTrackers,
            },
            mocks: { $store: store },
        });
    }

    describe("mounted()", () => {
        it("on init, the projects will be loaded", async () => {
            const loadProjects = jest
                .spyOn(rest_querier, "getSortedProjectsIAmMemberOf")
                .mockImplementation(() => Promise.resolve([]));

            await instantiateComponent();

            expect(loadProjects).toHaveBeenCalled();
        });
    });

    describe("loadProjects()", () => {
        beforeEach(() => {
            jest.spyOn(rest_querier, "getTrackersOfProject").mockImplementation(() =>
                Promise.resolve([])
            );
        });

        it("Displays an error when rest route fail", async () => {
            jest.spyOn(project_cache, "getSortedProjectsIAmMemberOf").mockRejectedValue([]);

            const wrapper = await instantiateComponent();
            await wrapper.vm.$nextTick(); // for the promise
            await wrapper.vm.$nextTick(); // for the finally

            expect(store.commit).toHaveBeenCalledWith("setErrorMessage", expect.anything());

            expect(wrapper.find("[data-test=tracker-loader]").exists()).toBe(false);
        });

        it("Displays the projects in selectbox", async () => {
            const first_project = { id: 543, label: "unheroically" } as Project;
            const second_project = { id: 544, label: "cycler" } as Project;

            jest.spyOn(project_cache, "getSortedProjectsIAmMemberOf").mockResolvedValue([
                first_project,
                second_project,
            ]);

            const wrapper = await instantiateComponent();
            await wrapper.vm.$nextTick();

            expect(wrapper.element).toMatchSnapshot();
            expect(wrapper.vm.$data.selected_project).toBe(first_project);
            expect(wrapper.vm.$data.projects).toEqual([first_project, second_project]);
        });
    });

    describe("loadTrackers()", () => {
        it("when I load trackers, the loader will be shown and the trackers options will be disabled if already selected", async () => {
            const project = { id: 543, label: "unheroically" } as Project;

            jest.spyOn(project_cache, "getSortedProjectsIAmMemberOf").mockResolvedValue([project]);

            const first_tracker = { id: 8, label: "coquettish" } as Tracker;
            const second_tracker = { id: 26, label: "unfruitfully" } as Tracker;
            const trackers = [first_tracker, second_tracker];
            jest.spyOn(rest_querier, "getTrackersOfProject").mockResolvedValue(trackers);
            const wrapper = await instantiateComponent([{ tracker_id: 26 } as SelectedTracker]);

            await wrapper.vm.$nextTick(); // for the promise of project
            await wrapper.vm.$nextTick(); // for the promise of tracker
            await wrapper.vm.$nextTick(); // for the finally

            expect(wrapper.vm.$data.trackers).toEqual(trackers);
        });

        it("when there is a REST error, it will be displayed", async () => {
            const project = { id: 543, label: "unheroically" } as Project;
            jest.spyOn(project_cache, "getSortedProjectsIAmMemberOf").mockResolvedValue([project]);
            jest.spyOn(rest_querier, "getTrackersOfProject").mockRejectedValue([]);

            const wrapper = await instantiateComponent();
            await wrapper.vm.$nextTick(); // for the promise of project
            await wrapper.vm.$nextTick(); // for the promise of tracker
            await wrapper.vm.$nextTick(); // for the finally

            expect(store.commit).toHaveBeenCalledWith("setErrorMessage", expect.anything());
        });
    });

    describe("addTrackerToSelection()", () => {
        it("when I add a tracker, then an event will be emitted", async () => {
            const project = { id: 972, label: "unheroically" } as Project;
            const selected_project = { id: 543, label: "unmortised" } as Project;
            jest.spyOn(project_cache, "getSortedProjectsIAmMemberOf").mockResolvedValue([
                project,
                selected_project,
            ]);

            const tracker = { id: 96, label: "simplus" } as Tracker;
            const selected_tracker = { id: 97, label: "acinus" } as Tracker;
            jest.spyOn(rest_querier, "getTrackersOfProject").mockResolvedValue([
                tracker,
                selected_tracker,
            ]);

            const wrapper = await instantiateComponent();

            await wrapper.vm.$nextTick(); // for the promise of project
            await wrapper.vm.$nextTick(); // for the promise of tracker
            await wrapper.vm.$nextTick(); // for the finally

            wrapper.vm.$data.selected_project = selected_project;
            wrapper.vm.$data.selected_tracker = selected_tracker;

            wrapper
                .find("[data-test=cross-tracker-selector-tracker-button]")
                .element.removeAttribute("disabled");
            wrapper.get("[data-test=cross-tracker-selector-tracker-button]").trigger("click");

            const emitted = wrapper.emitted()["tracker-added"];
            if (!emitted) {
                throw new Error("Event has not been emitted");
            }
            expect(emitted[0][0]).toEqual({
                selected_project,
                selected_tracker,
            });
            expect(wrapper.vm.$data.selected_tracker).toBe(null);
        });
    });
});
