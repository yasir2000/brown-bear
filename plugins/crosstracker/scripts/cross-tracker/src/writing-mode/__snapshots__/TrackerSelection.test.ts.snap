// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackerSelection loadProjects() Displays the projects in selectbox 1`] = `
<div
  class="cross-tracker-selector"
>
  <div
    class="tlp-form-element tlp-form-element-disabled"
    data-test="cross-tracker-selector-global"
  >
    <label
      class="tlp-label"
      for="project"
    >
      <translate-stub
        tag="span"
      >
        Project
      </translate-stub>
       
      <i
        class="fa fa-asterisk"
      />
    </label>
     
    <select
      class="cross-tracker-selector-project-input tlp-select"
      data-test="cross-tracker-selector-project"
      disabled="disabled"
      id="project"
      name="project"
    />
  </div>
   
  <div
    class="tlp-form-element tlp-form-element-disabled"
  >
    <label
      class="tlp-label"
      for="tracker"
    >
      <translate-stub
        tag="span"
      >
        Tracker
      </translate-stub>
       
      <i
        class="fa fa-asterisk"
      />
    </label>
     
    <div
      class="tlp-form-element tlp-form-element-append"
    >
      <select
        class="cross-tracker-selector-tracker-input tlp-select"
        data-test="cross-tracker-selector-tracker"
        disabled="disabled"
        id="tracker"
        name="tracker"
      >
        <option
          class="cross-tracker-please-choose-option"
          data-current-language="en_US"
          data-msgid="
                    Please choose...
                "
          value=""
        >
          
                    Please choose...
                
        </option>
         
      </select>
       
      <button
        class="tlp-append tlp-button-primary tlp-button-outline"
        data-test="cross-tracker-selector-tracker-button"
        disabled="disabled"
        type="button"
      >
        <i
          class="tlp-button-icon fas fa-circle-notch fa-spin"
          data-test="tracker-loader"
        />
         
        <translate-stub
          tag="span"
        >
          Add
        </translate-stub>
      </button>
    </div>
  </div>
</div>
`;
